src/twisted/internet/defer.py

```diff
@@ -159,6 +165,65 @@ def maybeDeferred(f, *args, **kw):
         return succeed(result)
 
 
+def unbox(deferred, raises=True, passthrough=True):
+    """
+    Unbox a L{Deferred} that has already been fully processed.
+
+    Take an already processed (i.e. no longer waiting on any callbacks to be
+    fired) and unbox it. If the result is a L{Failure}, then it will re-raise
+    the original exception, otherwise it will return the result of the
+    L{Deferred}.
+
+    B{This function CANNOT take a Deferred with pending asynchronous work to be
+    done and make it synchronous. It can ONLY unbox a Deferred which already
+    has successfully completed.}
+
+    @type deferred: Deferred
+    @param deferred: The Deferred to unbox.
+
+    @type raises: bool
+    @param raises: Whether or not to raise the original exception (L{True}) or
+        to return the failure object (L{False}).
+
+    @type passthrough: bool
+    @param passthrough: Whether or not to pass through the original value
+        (L{True}) or to consume any results, error or otherwise (L{False}).
+
+    @raise InvalidStateError: If C{deferred} has pending work to be done and
+        has no result to unbox.
+
+    @return: The result of the L{Deferred}.
+    """
+    results = []
+
+    @deferred.addBoth
+    def done(result):
+        results[:] = [result]
+        # We return the result here so that if anything else continued to use
+        # this Deferred after it's been unboxed then the act of unboxing
+        # has not broken the flow of values through the callback chain.
+        if passthrough:
+            return result
+
+    # If we've gotten to this point and we haven't had our collect function
+    # called, then we must be waiting on something in the Deferred chain. If so
+    # we'll error out here because there's nothing we can do to advance the
+    # state of the Deferred. If we've got a result then we'll go ahead and
+    # return it or raise the given exception.
+    if not results:
+        raise InvalidStateError(
+            "Cannot call unbox with a Deferred that has pending work to be "
+            "done."
+        )
+    else:
+        [result] = results
+
+        if raises and isinstance(result, failure.Failure):
```

Comment:
I think you'll run into trouble here if you have a function that callbacks with a failure (but maybe that will already fail in a place you don't control).

But like, doing:

`result, failure = [], []`

and then `addCallback` / `addErrback` would let you distinguish "properly".

Also I think you need tests both for this specifically and for the `BaseException` vs. `Exception` distinction (unless I missed an existing one for the latter).